Use `Vector` for intermediate computations in Dijkstra

IvanIsCoding opened the pull request:
<!--
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I ran rustfmt locally
- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

Related to #492 

Replaces `DictMap` with `Vector` for intermediate computations in our Dijkstra's implementation.

* Instead of always using `DictMap` to store the scores, we use a `Vector` and create the `DictMap` the end. That way, users can get a nice deterministic output and we can keep computations fast

mtreinish commented:
This makes the overhead worse, adding a sort step looks like it causes an even worse regression:

![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/143297186-03741af6-a080-4dfd-a214-433a08be3c3c.png)
![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/143297187-a8a7f444-9134-434b-85c6-21a95e93fa60.png)
![all_pairs](https://user-images.githubusercontent.com/2447371/143297188-02a6b011-f1b6-4805-9b2d-2b82419c5287.png)

IvanIsCoding commented:
> This makes the overhead worse, adding a sort step looks like it causes an even worse regression:
> 
> ![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/143297186-03741af6-a080-4dfd-a214-433a08be3c3c.png) ![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/143297187-a8a7f444-9134-434b-85c6-21a95e93fa60.png) ![all_pairs](https://user-images.githubusercontent.com/2447371/143297188-02a6b011-f1b6-4805-9b2d-2b82419c5287.png)

Good to know, I will try another idea

IvanIsCoding commented:
Tried using `Vector` instead of `HashMap`, that might give us more margin.

georgios-ts commented on retworkx-core/src/shortest_path/dijkstra.rs:
```diff
@@ -169,5 +170,11 @@ where
     }
-    Ok(scores)
+
+    Ok(graph
+        .node_identifiers()
+        .filter_map(|node_id| {
+            scores[graph.to_index(node_id)].map(|score| (node_id, score))
+        })
+        .collect())
```
Iterating directly over scores should be faster.
```suggestion
    Ok(scores
        .into_iter()
        .enumerate()
        .filter_map(|(node, score)| {
            score.map(|val| (graph.from_index(node), val))
        })
        .collect())
```

georgios-ts reviewed (commented):
While we are here, we could also see if there is any performance benfit replacing `BinaryHeap` with a [`QuaternaryHeap`](https://docs.rs/dary_heap/0.3.2/dary_heap/type.QuaternaryHeap.html). At least, BGL (that graph-tool depends on) uses a 4-ary heap.

IvanIsCoding commented:
> While we are here, we could also see if there is any performance benfit replacing `BinaryHeap` with a [`QuaternaryHeap`](https://docs.rs/dary_heap/0.3.2/dary_heap/type.QuaternaryHeap.html). At least, BGL (that graph-tool depends on) uses a 4-ary heap.

That crate seems a bit too young, I will hold on the Quaternary Heap for a bit

IvanIsCoding commented:
I added the QuaternaryHeap, they had a version without const generics with Rust MSRV 1.31

IvanIsCoding commented:
@mtreinish I think the biggest culprit is https://github.com/Qiskit/retworkx/blob/main/src/shortest_path/mod.rs#L221-L231, every time we map through a HashMap/DictMap we allocate more memory. Because DictMap uses more memory and the graph had millions of nodes, the results got worse.

The code I have now is not the cleanest, but it creates a DictMap only once. That does lower the performance regression

mtreinish commented:
@IvanIsCoding yeah that does the trick now. Running the benchmarks again it's faster than 0.10.2 now:

![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/144129155-c91a3c07-3240-40bb-90ba-82d11dc40d4b.png)
![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/144129158-e04e8528-4700-426e-b015-602cf656bccc.png)
![all_pairs](https://user-images.githubusercontent.com/2447371/144129160-5e50d25b-278a-499e-9c1b-eaaf32443603.png)

IvanIsCoding commented:
I will try to clean the code for this one, and flag other similar cases in #492.

georgios-ts commented:
It's great that we managed to fixed the regression! Having two separate functions `dijkstra`, `dijkstra_vector` is not the end of the world but it's not the cleanest interface, so I have some guesses/suggestions to make:

1. Given that the biggest regression was in the single source shortest path benchmark between two nodes and hopefully the code https://github.com/Qiskit/retworkx/blob/main/src/shortest_path/mod.rs#L221-L231 does not allocate memory for million of nodes if `goal` is set, I'm wondering if the early return here https://github.com/Qiskit/retworkx/blob/4673692bff01f924f71c6aeba8e8203db7fea6ee/src/shortest_path/mod.rs#L223-L234 that avoids iterating over all nodes in the dict could fix alone the regression.
2. Keep the code from main in `dijkstra` and change the type of `scores` from `DictMap<G::NodeId, K>` to `DictMap<usize, K>`, so we avoid iterate/and allocate a new dict to store inside `PathLengthMapping`. Or, do the opposite and change the dict type inside `PathLengthMapping` to `DictMap<NodeIndex, f64>`.
3. Keep only `dijkstra_vector` in retworkx-core.
4. Define a trait:
    ```rust
   pub trait DistMap<N, K> { 
       fn get(&self,  a: N) -> Option<K>;
       fn put(&mut self, a: N, val: K)
   }
    ```
    similar to [`VisitMap`](https://docs.rs/petgraph/latest/petgraph/visit/trait.VisitMap.html), implement it for `HashMap`/`DictMap`/`Vec` and make `dijkstra` generic like

    ```rust
   pub fn dijkstra<G, F, K, E>(
        graph: G,
        start: G::NodeId,
        goal: Option<G::NodeId>,
        edge_cost: F,
        scores: impl DistMap<G::NodeId, K>
        path: Option<&mut DictMap<G::NodeId, Vec<G::NodeId>>>,
    )
    ```
    so Rust users can choose between `HashMap`/`DictMap`/`Vec`.

IvanIsCoding commented:
> It's great that we managed to fixed the regression! Having two separate functions `dijkstra`, `dijkstra_vector` is not the end of the world but it's not the cleanest interface, so I have some guesses/suggestions to make:
> 
> 1. Given that the biggest regression was in the single source shortest path benchmark between two nodes and hopefully the code https://github.com/Qiskit/retworkx/blob/main/src/shortest_path/mod.rs#L221-L231 does not allocate memory for million of nodes if `goal` is set, I'm wondering if the early return here https://github.com/Qiskit/retworkx/blob/4673692bff01f924f71c6aeba8e8203db7fea6ee/src/shortest_path/mod.rs#L223-L234
>     that avoids iterating over all nodes in the dict could fix alone the regression.
> 2. Keep the code from main in `dijkstra` and change the type of `scores` from `DictMap<G::NodeId, K>` to `DictMap<usize, K>`, so we avoid iterate/and allocate a new dict to store inside `PathLengthMapping`. Or, do the opposite and change the dict type inside `PathLengthMapping` to `DictMap<NodeIndex, f64>`.
> 3. Keep only `dijkstra_vector` in retworkx-core.
> 4. Define a trait:
>    ```rust
>    pub trait DistMap<N, K> { 
>       fn get(&self,  a: N) -> Option<K>;
>       fn put(&mut self, a: N, val: K)
>    }
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    similar to [`VisitMap`](https://docs.rs/petgraph/latest/petgraph/visit/trait.VisitMap.html), implement it for `HashMap`/`DictMap`/`Vec` and make `dijkstra` generic like
>    ```rust
>    pub fn dijkstra<G, F, K, E>(
>        graph: G,
>        start: G::NodeId,
>        goal: Option<G::NodeId>,
>        edge_cost: F,
>        scores: impl DistMap<G::NodeId, K>
>        path: Option<&mut DictMap<G::NodeId, Vec<G::NodeId>>>,
>    )
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    so Rust users can choose between `HashMap`/`DictMap`/`Vec`.

I like the trait idea, I will try implementing it. It would be nice to have a flexible API that works with all DictMap/HashMap/Vec

IvanIsCoding commented:
We will need to benchmark again, but I refactored the code such that the function in retworkx-core can return both a Vec and an IndexMap

mtreinish commented:
> We will need to benchmark again, but I refactored the code such that the function in retworkx-core can return both a Vec and an IndexMap

![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/144628925-d05d827d-4e3f-4966-ad9a-7d7d60ba6984.png)
![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/144628928-9ae796dc-1a02-4288-b32f-f1c14b478c24.png)
![all_pairs](https://user-images.githubusercontent.com/2447371/144628929-650890cc-f063-497f-a231-a0a20147c41f.png)

georgios-ts commented on retworkx-core/src/shortest_path/dijkstra.rs:
```diff
@@ -20,14 +20,54 @@
 use crate::min_scored::MinScored;
 
+// Trait for supporting storing
+pub trait DistanceMap<K, V> {
```
Probably it's better to move this into a separate file since we might need it for other shortest path algorithms too. Also we might need to add documentation on this trait since it'll be in the docs of `retworkx-core` crate. We can even give it another name since we can in fact store any property and not just distances.

georgios-ts commented on retworkx-core/src/shortest_path/dijkstra.rs:
```diff
@@ -20,14 +20,54 @@
 
+// Trait for supporting storing
+pub trait DistanceMap<K, V> {
+    fn new_distancemap(num_elements: usize) -> Self;
```
I'd prefer naming this method like
```suggestion
    fn distance_map(num_elements: usize) -> Self;
```
or just `new` or `build`

georgios-ts commented on retworkx-core/src/shortest_path/dijkstra.rs:
```diff
@@ -20,14 +20,54 @@
+    }
+}
+
+impl<K: Eq + Hash, V: Clone> DistanceMap<K, V> for DictMap<K, V> {
```
We can add a similar implementation for `HashMap`.

georgios-ts commented on retworkx-core/src/shortest_path/dijkstra.rs:
````diff
@@ -98,24 +138,25 @@ use crate::min_scored::MinScored;
 /// // z is not inside res because there is not path from b to z.
 /// ```
-pub fn dijkstra<G, F, K, E>(
+pub fn dijkstra<G, F, K, E, S>(
````
Can we mention in the docs that you can choose the return type for this function? It currently states that it returns a `DictMap`.

georgios-ts commented on retworkx-core/src/shortest_path/dijkstra.rs:
```diff
@@ -98,24 +138,25 @@ use crate::min_scored::MinScored;
-    G: IntoEdges + Visitable,
-    G::NodeId: Eq + Hash,
+    G: IntoEdges + Visitable + NodeIndexable,
+    G::NodeId: Eq + Hash + IndexType,
```
I don't think this is needed and actually it'd restrict the graph types that we can use.

```suggestion
    G::NodeId: Eq + Hash,
```

georgios-ts commented on retworkx-core/src/shortest_path/k_shortest_path.rs:
```diff
@@ -83,13 +84,13 @@ pub fn k_shortest_path<G, F, E, K>(
 {
     let mut counter: Vec<usize> = vec![0; graph.node_bound()];
-    let mut scores = DictMap::with_capacity(graph.node_count());
+    let mut scores: Vec<Option<K>> = vec![None; graph.node_bound()];
```
We should also use here `DistanceMap`  for maximum flexibility, right?

georgios-ts commented on src/shortest_path/all_pairs_dijkstra.rs:
```diff
@@ -81,23 +81,24 @@ pub fn all_pairs_dijkstra_path_lengths<Ty: EdgeType + Sync>(
-                    }
-                })
-                .collect(),
+                    as PyResult<DictMap<NodeIndex, f64>>)
```
Type casting here doesn't look great (in terms of readability). IMO, type annotation is nicer:

```rust
let path_lengths: PyResult<DictMap<NodeIndex, f64>> =
     dijkstra(graph, x, None, |e| edge_cost(e.id()), None);

let out_map = PathLengthMapping {
     path_lengths: path_lengths
            .unwrap()
            .iter()
            ...
            .collect(),
};
```

georgios-ts reviewed (commented):
I left some inline comments. Apart from that, it's weird that we use internally a vector for the computation in `dijkstra_shortest_path_lengths` but a `DictMap` in `dijkstra_shortest_paths`. For maximum flexibility and control over memory - speed trade-off, it seems a good choice to add a kwarg `sparse` in order to let users choose which data structure best fits their needs.

IvanIsCoding commented:
> I left some inline comments. Apart from that, it's weird that we use internally a vector for the computation in `dijkstra_shortest_path_lengths` but a `DictMap` in `dijkstra_shortest_paths`. For maximum flexibility and control over memory - speed trade-off, it seems a good choice to add a kwarg `sparse` in order to let users choose which data structure best fits their needs.

I addressed the comments, but I think adding a `sparse` kwarg is not the best choice. We'd need to have a conditional calling the same function with different types, which is a bit cumbersome. I think it's ok to make a choice for the user in this case.

georgios-ts commented:
Yeah, it's ok to make a choice for the users but then why not use a vector internally for `dijkstra_shortest_paths` (or even in `all_pairs_dijkstra_path_lengths`) too?

IvanIsCoding commented:
> all_pairs_dijkstra_path_lengths

I've updated the PR to include that as well

mtreinish commented on retworkx-core/src/distancemap.rs:
```diff
@@ -0,0 +1,88 @@
+use crate::dictmap::*;
+use hashbrown::HashMap;
+
+pub trait DistanceMap<K, V> {
```
Do we need a docstring for this trait? I'm not sure module docs are sufficient to document how this is used or what the expectations for it are.

mtreinish replied:
I took a look at the output for cargo doc, I think we should add a docstring here. When you click `DistanceMap` from the shortest path functions you get this page:

![Screenshot_2022-01-14_13-03-00](https://user-images.githubusercontent.com/2447371/149563526-28966b27-53d1-4125-9ac9-18a29177c906.png)

The module page is visible if you explicitly go to it, but without manually searching for it you probably won't see it. So adding a docstring explaining what this trait is used for and what the methods in the trait do is probably good (just in case someone wants to implement it for their own type).

mtreinish commented on retworkx-core/src/distancemap.rs:
```diff
@@ -0,0 +1,88 @@
+// License for the specific language governing permissions and limitations
+// under the License.
+
+// This module was originally copied and forked from the upstream petgraph
+// repository, specifically:
+// https://github.com/petgraph/petgraph/blob/0.5.1/src/dijkstra.rs
+// this was necessary to modify the error handling to allow python callables
+// to be use for the input functions for edge_cost and return any exceptions
+// raised in Python instead of panicking
```
Is this true or is it just a copy and paste error.

IvanIsCoding replied:
This was a copy and paste error

mtreinish reviewed (approved):
Overall LGTM, thanks for keeping with this. The code looks great and the performance definitely is where we want it to be:

![all_pairs](https://user-images.githubusercontent.com/2447371/149423328-54fb925c-94ff-4942-91bc-fe9830daf55f.png)
![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/149423329-92e19c74-e33c-4681-83f1-c47b9ee3686b.png)
![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/149423331-73e073ae-81f1-45cd-8cb9-6c7610fdd15d.png)

It'd be good if @georgios-ts gave this a look too since he had comments on earlier revisions.

I just had two quick comments inline about the documentation and comments before we merge this around the `DistanceMap` trait. It might be worth manually checking the `cargo doc` output to verify how it renders.

georgios-ts reviewed (approved):
LGTM and performance looks great!
